sdk/physical/cache.go

```diff
@@ -197,14 +230,69 @@ func (c *Cache) Delete(ctx context.Context, key string) error {
 	return err
 }
 
-func (c *Cache) List(ctx context.Context, prefix string) ([]string, error) {
+func (c *cache) List(ctx context.Context, prefix string) ([]string, error) {
 	// Always pass-through as this would be difficult to cache. For the same
 	// reason we don't lock as we can't reasonably know which locks to readlock
 	// ahead of time.
 	return c.backend.List(ctx, prefix)
 }
 
-func (c *Cache) ListPage(ctx context.Context, prefix string, after string, limit int) ([]string, error) {
+func (c *cache) ListPage(ctx context.Context, prefix string, after string, limit int) ([]string, error) {
 	// See note above about List(...).
 	return c.backend.ListPage(ctx, prefix, after, limit)
 }
+
+func (c *cache) cloneWithStorage(b Backend) *cache {
+	// We construct a new cache here: this starts the transaction with a
+	// fresh, localized cache. This is globally sub-optimal (as it purges
+	// everything), but easiest to implement (as the transaction can modify
+	// its cache as it pleases).
+	return NewCache(b, c.size, c.logger, c.metricSink).(*cache)
+}
+
+func (c *transactionalCache) BeginReadOnlyTx(ctx context.Context) (Transaction, error) {
+	txn, err := c.cache.backend.(TransactionalBackend).BeginReadOnlyTx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// txn does not implement TransactionalBackend because we don't support
+	// nested transactions so this will always cast to *cache.
+	ctxn := c.cache.cloneWithStorage(txn)
+	return &cacheTransaction{
+		*ctxn,
+		c,
+	}, nil
+}
+
+func (c *transactionalCache) BeginTx(ctx context.Context) (Transaction, error) {
+	txn, err := c.cache.backend.(TransactionalBackend).BeginTx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// See note above in BeginReadOnlyTx(...).
+	ctxn := c.cache.cloneWithStorage(txn)
+	return &cacheTransaction{
+		*ctxn,
+		c,
+	}, nil
+}
+
+func (c *cacheTransaction) Commit(ctx context.Context) error {
+	if err := c.cache.backend.(Transaction).Commit(ctx); err != nil {
+		return err
+	}
+
+	c.parent.Purge(ctx)
```

Comment:
I think this warrants a comment: we need to purge the cache when we commit the transaction as otherwise we'd need to do the handling in `Put(...)` or `Delete(...)` and replay write operations and invalidate their entries in the parent.

Technically this isn't necessary for `Rollback` since no data was modified so I'll remove this below.

@JanMa do you agree?

Reply:
I ended up strictly doing invalidations of modified entries. This has also been pushed to #292 so that we can ensure it passes the new transactional test suite :)